File: Roboti/app/src/main/java/cz/muni/fi/umimecesky/roboti/utils/BotLogicSlow.java
package cz.muni.fi.umimecesky.roboti.utils;

import android.util.Log;

import cz.muni.fi.umimecesky.roboti.pojo.RaceConcept;

class BotLogicSlow implements RobotLogic {

    private RaceConcept concept;

    public BotLogicSlow(RaceConcept concept) {
        this.concept = concept;
        Log.v("slow concept", String.valueOf(concept.levelProgress()));
        Log.v("slow is before half", String.valueOf(isBeforeHalf()));
    }

    @Override
    public int millisecondsPerSolution() {
        return 3500 - (int) (1500 * concept.levelProgress());
    }

    @Override
    public double correctnessRatio() {
        double addition = isBeforeHalf() ? Utils.roundBy2(0.4 * concept.levelProgress()) :
                0.1 * concept.levelProgress();
        return 0.75 + addition;
    }

    @Override
    public int hopsPerCorrect() {
        return isBeforeHalf() ? 1 : 2;
    }

    private boolean isBeforeHalf() {
        return concept.levelProgress() < 0.5;
    }
}

File: app/src/main/java/cz/muni/fi/umimecesky/game/robots/logic/BotLogicSlow.kt
/*
 * Copyright (c) 2018 Marek Sabo
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *                http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cz.muni.fi.umimecesky.game.robots.logic

import cz.muni.fi.umimecesky.game.robots.logic.RobotLogic.Companion.roundBy2places
import cz.muni.fi.umimecesky.game.shared.model.RaceConcept

class BotLogicSlow(private val concept: RaceConcept) : RobotLogic {

    override fun millisecondsPerSolution(): Long = 3500L - (1500L * concept.levelProgress()).toLong()

    override fun correctnessRatio(): Double {
        val addition = if (isBeforeHalf)
            roundBy2places(0.4 * concept.levelProgress())
        else
            0.1 * concept.levelProgress()
        return 0.75 + addition
    }

    override fun hopsPerCorrect(): Int = if (isBeforeHalf) 1 else 2

    private val isBeforeHalf: Boolean
        get() = concept.levelProgress() < 0.5
}
